src/cow-react/modules/tokens/hooks/useOnchainBalances.ts

```diff
@@ -0,0 +1,144 @@
+const DEFAULT_LISTENER_OPTIONS: ListenerOptionsWithGas = { gasRequired: 185_000, blocksPerFetch: 5 }
+
+export type OnchainTokenAmount = OnchainState<CurrencyAmount<Token> | undefined>
+
+export type OnchainTokenAmounts = {
+  [tokenAddress: string]: OnchainTokenAmount
+}
+
+export type OnchainAmountsResult = {
+  amounts: OnchainTokenAmounts
+  isLoading: boolean
+}
+
+export interface OnchainAmountsParams {
+  account?: string
+  tokens?: (Token | undefined)[]
+  blocksPerFetch?: number
+}
+
+export type OnchainBalancesParams = OnchainAmountsParams
+
+export type OnchainAllowancesParams = OnchainAmountsParams & { spender?: string }
+
+export function useOnchainBalances(params: OnchainBalancesParams): OnchainAmountsResult {
+  const { account } = params
+  const callParams = useMemo(() => [account], [account])
+  return useOnchainErc20Amounts('balanceOf', callParams, params)
+}
+
+export function useOnchainAllowances(params: OnchainAllowancesParams): OnchainAmountsResult {
+  const { account, spender } = params
+  const callParams = useMemo(() => [account, spender], [account, spender])
+  return useOnchainErc20Amounts('allowance', callParams, params)
+}
+
+export interface OnchainBalancesAndAllowancesParams {
+  account: string | undefined
+  spender: string | undefined
+  tokens: Token[]
+  blocksPerFetchBalance?: number
+  blocksPerFetchAllowance?: number
+}
+
+export interface OnchainBalancesAndAllowances {
+  balances: OnchainTokenAmounts
+  allowances: OnchainTokenAmounts
+  isLoading: boolean
+}
+
+/**
+ * Fetches
+ * @param params
+ * @returns
+ */
+export function useOnchainBalancesAndAllowances(
+  params: OnchainBalancesAndAllowancesParams
+): OnchainBalancesAndAllowances {
+  const { account, spender, tokens, blocksPerFetchAllowance, blocksPerFetchBalance } = params
+
+  const { amounts: balances, isLoading: areBalancesLoading } = useOnchainBalances({
+    account,
+    tokens,
+    blocksPerFetch: blocksPerFetchBalance,
+  })
+  const { amounts: allowances, isLoading: areAllowancesLoading } = useOnchainAllowances({
+    account,
+    tokens,
+    spender,
+    blocksPerFetch: blocksPerFetchAllowance,
+  })
+
+  return useMemo(
+    () => ({
+      balances,
+      allowances,
+      isLoading: areBalancesLoading || areAllowancesLoading,
+    }),
+    [balances, allowances, areBalancesLoading, areAllowancesLoading]
+  )
+}
+
+function useOnchainErc20Amounts(
+  erc20Method: 'balanceOf' | 'allowance',
+  callParams: (string | undefined)[],
+  params: OnchainAmountsParams
+): OnchainAmountsResult {
+  const { account, blocksPerFetch, tokens } = params
+
+  const validatedTokens: Token[] = useMemo(
+    () => tokens?.filter((t?: Token): t is Token => isAddress(t?.address) !== false) ?? [],
+    [tokens]
+  )
+  const validatedTokenAddresses = useMemo(() => validatedTokens.map((vt) => vt.address), [validatedTokens])
+
+  // Do on-chain calls
+  const balancesCallState = useMultipleContractSingleData(
```

Comment:
https://github.com/cowprotocol/cowswap/pull/2306

@anxolin @alfetopito I don't propose it, just an example.
I would like to have an opportunity to use multicall as a function, not as a hook.

Reply:
Ohh, I did that in the past.
I guess it's somewhere around in the code